test/extensions/filters/http/buffer/buffer_filter_integration_test.cc

```diff
@@ -63,6 +63,34 @@ TEST_P(BufferIntegrationTest, RouterRequestPopulateContentLength) {
   EXPECT_EQ("200", response->headers().Status()->value().getStringView());
 }
 
+TEST_P(BufferIntegrationTest, RouterRequestPopulateContentLengthOnTrailers) {
+  config_helper_.addFilter(ConfigHelper::DEFAULT_BUFFER_FILTER);
+  initialize();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto encoder_decoder = codec_client_->startRequest(Http::TestHeaderMapImpl{
+      {":method", "POST"}, {":scheme", "http"}, {":path", "/shelf"}, {":authority", "host"}});
+  request_encoder_ = &encoder_decoder.first;
+  IntegrationStreamDecoderPtr response = std::move(encoder_decoder.second);
+  codec_client_->sendData(*request_encoder_, "0123", false);
+  codec_client_->sendData(*request_encoder_, "456", false);
+  codec_client_->sendData(*request_encoder_, "789", false);
+  codec_client_->sendTrailers(*request_encoder_, Http::TestHeaderMapImpl{});
```

Comment:
Trying to improve my knowledge of internal envoy logic: is this semantically different from `sendData(..., true)`?
How are either of them materialized on the wire?

Reply:
yep it is different. `sendTrailers` will encode a HEADERS frame with END_STREAM true while `sendData(data, true)` encodes a DATA frame with END_STREAM true. Envoy mobile closes client streams with trailers, so the content-length headers was not getting set in our edge infrastructure.